File: backend/src/utils/http.ts
import type { Response } from 'express';

export function ok<T, M = Record<string, never>>(res: Response, data: T, meta?: M): Response {
  return res.json({
    success: true,
    data,
    ...(meta ? { meta } : {}),
  });
}

export function fail(
  res: Response,
  status: number,
  code: string,
  message: string,
  details?: unknown
): Response {
  return res.status(status).json({
    success: false,
    error: {
      code,
      message,
      ...(details !== undefined ? { details } : {}),
    },
  });
}

File: backend/app/utils/http.py
from __future__ import annotations

from typing import Any

from fastapi.responses import JSONResponse


def success_payload(data: Any, meta: Any | None = None) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "success": True,
        "data": data,
    }
    if meta is not None:
        payload["meta"] = meta
    return payload


def error_payload(code: str, message: str, details: Any | None = None) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "success": False,
        "error": {
            "code": code,
            "message": message,
        },
    }
    if details is not None:
        payload["error"]["details"] = details
    return payload


def error_response(
    status_code: int,
    code: str,
    message: str,
    details: Any | None = None,
) -> JSONResponse:
    return JSONResponse(
        status_code=status_code,
        content=error_payload(code, message, details),
    )
